File: app/src/main/java/cz/johnyapps/cheers/adapters/CountersAdapter.java
package cz.johnyapps.cheers.adapters;

import android.content.Context;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.ArrayList;
import java.util.List;

import cz.johnyapps.cheers.R;
import cz.johnyapps.cheers.entities.CounterWithBeverage;
import cz.johnyapps.cheers.views.CounterView;

public class CountersAdapter extends SelectableAdapter<CountersAdapter.CounterViewHolder, CounterWithBeverage, List<CounterWithBeverage>> {
    @NonNull
    private List<CounterWithBeverage> countersWithBeverages;

    public CountersAdapter(@NonNull Context context,
                           @Nullable List<CounterWithBeverage> countersWithBeverages,
                           @NonNull OnSelectListener<CounterWithBeverage> onSelectListener) {
        super(context, onSelectListener);
        this.countersWithBeverages = countersWithBeverages == null ? new ArrayList<>() : countersWithBeverages;
    }

    @NonNull
    @Override
    public CounterWithBeverage getItem(int position) {
        return countersWithBeverages.get(position);
    }

    @Override
    protected void onUpdate(@Nullable List<CounterWithBeverage> countersWithBeverages) {
        this.countersWithBeverages = countersWithBeverages == null ? new ArrayList<>() : countersWithBeverages;
    }

    @NonNull
    @Override
    public CounterViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new CounterViewHolder(getInflater().inflate(R.layout.item_counter, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull CounterViewHolder holder, int position, boolean selected) {
        CounterWithBeverage counterWithBeverage = countersWithBeverages.get(position);
        holder.counterView.setCounter(counterWithBeverage);
        holder.counterView.setOnPassClickListener(v -> selectPosition(position));
        holder.counterView.setPassClicks(selected);
    }

    @Override
    public int getItemCount() {
        return countersWithBeverages.size();
    }

    public class CounterViewHolder extends SelectableAdapter<CounterViewHolder, CounterWithBeverage, List<CounterWithBeverage>>.SelectableViewHolder {
        CounterView counterView;

        public CounterViewHolder(@NonNull View root) {
            super(root);
            counterView = root.findViewById(R.id.counterView);
        }
    }
}

File: app/src/main/java/cz/johnyapps/cheers/adapters/CountersAdapter.kt
package cz.johnyapps.cheers.adapters

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.DiffUtil
import cz.johnyapps.cheers.R
import cz.johnyapps.cheers.databinding.ItemCounterBinding
import cz.johnyapps.cheers.entities.CounterWithBeverage

class CountersAdapter: SelectableAdapter<CounterWithBeverage, CountersAdapter.ViewHolder>(DIFF_CALLBACK) {
    var allCountersDisabled = false
        set(value) {
            notifyDataSetChanged()
            field = value
        }
    var onCounterClickListener: OnCounterClickListener? = null

    init {
        setHasStableIds(true)
    }

    override fun getItemId(position: Int): Long {
        return getItem(position).counter.id
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int, selected: Boolean) {
        holder.binding.counterWithBeverage = getItem(position)
        holder.binding.counterView.setPassClicks(allCountersDisabled)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder(DataBindingUtil.inflate(LayoutInflater.from(parent.context),
            R.layout.item_counter,
            parent,
            false))
    }

    open inner class ViewHolder(val binding: ItemCounterBinding): SelectableAdapter<CounterWithBeverage, CountersAdapter.ViewHolder>.ViewHolder(binding.root) {
        init {
            addRootOnClickListener(object : OnRootClickListener {
                override fun onClick(view: View) {
                    if (allCountersDisabled) {
                        onCounterClickListener?.onClick(getItem(adapterPosition))
                    }
                }
            })
        }
    }

    companion object {
        private val DIFF_CALLBACK: DiffUtil.ItemCallback<CounterWithBeverage> = object: DiffUtil.ItemCallback<CounterWithBeverage>(){
            override fun areItemsTheSame(oldItem: CounterWithBeverage, newItem: CounterWithBeverage): Boolean {
                return oldItem.counter.id == newItem.counter.id
            }

            override fun areContentsTheSame(oldItem: CounterWithBeverage, newItem: CounterWithBeverage): Boolean {
                return oldItem.counter == newItem.counter
            }

        }
    }

    interface OnCounterClickListener {
        fun onClick(counterWithBeverage: CounterWithBeverage)
    }
}